tests/testsuite/build_script.rs

```diff
@@ -2207,6 +2215,206 @@ fn test_duplicate_deps() {
+        )
+        .file(
+            "src/main.rs",
+            r#"
+                extern crate bar;
+                fn main() { bar::do_nothing() }
+            "#,
+        )
+        .file(
+            "build.rs",
+            r#"
+                extern crate bar;
+                use std::env;
+                fn main() {
+                    bar::do_nothing();
+                    assert_eq!(env::var("DEP_FOO_FOO").unwrap(), "bar");
+                    assert_eq!(env::var("CARGO_BUILD_TYPE").unwrap(), "cross");
+                }
+            "#,
+        )
+        .file(
+            "bar/Cargo.toml",
+            r#"
+                [project]
+                name = "bar"
+                version = "0.1.0"
+                authors = []
+                links = "foo"
+                build = "build.rs"
+            "#,
+        )
+        .file(
+            "bar/build.rs",
+            r#"
+                use std::env;
+                fn main() {
+                    println!("cargo:foo=bar");
+                    assert_eq!(env::var("CARGO_BUILD_TYPE").unwrap(), "cross");
+                }
+            "#,
+        )
+        .file("bar/src/lib.rs", "pub fn do_nothing() {}")
+        .build();
+
+    p.cargo("build -v --target").arg(&target).run();
+}
+
+#[cargo_test]
+fn test_duplicate_shared_deps_alt_cross() {
+    if cross_compile::disabled() {
+        return;
+    }
+    let cross_target = cross_compile::alternate();
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "foo"
+                version = "0.1.0"
+                authors = []
+                build = "build.rs"
+
+                [dependencies]
+                bar = { path = 'bar' }
+
+                [build-dependencies]
+                bar = { path = 'bar' }
+            "#,
+        )
+        .file(
+            "src/main.rs",
+            r#"
+                extern crate bar;
+                fn main() { bar::do_nothing() }
+            "#,
+        )
+        .file(
+            "build.rs",
+            r#"
+                extern crate bar;
+                use std::env;
+                fn main() {
+                    bar::do_nothing();
+                    assert_eq!(env::var("DEP_FOO_FOO").unwrap(), "bar");
+                    assert_eq!(env::var("CARGO_BUILD_TYPE").unwrap(), "cross");
+                }
+            "#,
+        )
+        .file(
+            "bar/Cargo.toml",
+            r#"
+                [project]
+                name = "bar"
+                version = "0.1.0"
+                authors = []
+                links = "foo"
+                build = "build.rs"
+            "#,
+        )
+        .file(
+            "bar/build.rs",
+            r#"
+                use std::env;
+                fn main() {
+                    println!("cargo:foo=bar");
+                    assert_eq!(env::var("CARGO_BUILD_TYPE").unwrap(), "cross");
```

Comment:
I believe this is also a bug because this build script should be run twice, once with the host and once with the target.

Reply:
Hmm, any idea how I should reproduce this issue in a test so that it fails?

Reply:
You could have the build script produce different values based on `CARGO_BUILD_TYPE` and both of those could get fed into the final binary. The final binary could then print things out and assert that they're equal to their expected values.

Reply:
On further review this current behavior appears to be what is intended, `CARGO_BUILD_TYPE` is for indicating if a target flag was passed to cargo, it's not intended to be used for determining if a build script is a host or target dependency.

Reply:
Ok, so I think I managed to reproduce this issue while adding a `CARGO_BUILD_DEPENDENCY_TYPE` env variable and while attempting to filter `CARGO_BUILD_TYPE` to only be present for target dependency types.